File: src/scrapers/serper.py
"""
Serper.dev Client
Google Search API para buscas no Brasil
https://serper.dev/
"""

import re
from typing import Any, Dict, List, Optional

import structlog

from config.settings import settings

from .base import BaseScraper

logger = structlog.get_logger()


class SerperClient(BaseScraper):
    """
    Cliente para Serper.dev - Google Search API

    Funcionalidades:
    - Busca Google padrão
    - Busca de imagens
    - Busca de notícias
    - Busca de lugares
    - Autocomplete
    """

    # Metadados da fonte para rastreabilidade (CLAUDE.md)
    SOURCE_NAME = "Serper - Google Search"
    SOURCE_PROVIDER = "Serper.dev"
    SOURCE_CATEGORY = "api"
    SOURCE_COVERAGE = "Resultados de busca Google, notícias, imagens"
    SOURCE_DOC_URL = "https://serper.dev/docs"

    def __init__(
        self,
        api_key: Optional[str] = None,
        timeout: float = 30.0
    ):
        super().__init__(
            api_key=api_key or settings.serper_api_key,
            base_url="https://google.serper.dev",
            rate_limit=100,
            timeout=timeout
        )

    def _get_headers(self) -> Dict[str, str]:
        return {
            "X-API-KEY": self.api_key,
            "Content-Type": "application/json"
        }

    async def search(
        self,
        query: str,
        num: int = 10,
        gl: str = "br",
        hl: str = "pt-br",
        page: int = 1
    ) -> Dict[str, Any]:
        """
        Busca Google padrão

        Args:
            query: Termo de busca
            num: Número de resultados (max 100)
            gl: País (br = Brasil)
            hl: Idioma (pt-br = Português Brasil)
            page: Página de resultados

        Returns:
            Resultados da busca
        """
        logger.info("serper_search", query=query[:50])

        result = await self.post("/search", json={
            "q": query,
            "num": min(num, 100),
            "gl": gl,
            "hl": hl,
            "page": page
        })

        return {
            "query": query,
            "organic": result.get("organic", []),
            "knowledge_graph": result.get("knowledgeGraph"),
            "answer_box": result.get("answerBox"),
            "related_searches": result.get("relatedSearches", []),
            "people_also_ask": result.get("peopleAlsoAsk", []),
            "total_results": len(result.get("organic", []))
        }

    async def search_news(
        self,
        query: str,
        num: int = 10,
        gl: str = "br",
        hl: str = "pt-br",
        tbs: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Busca de notícias

        Args:
            query: Termo de busca
            num: Número de resultados
            gl: País
            hl: Idioma
            tbs: Filtro de tempo (qdr:d = último dia, qdr:w = última semana, qdr:m = último mês)

        Returns:
            Notícias encontradas
        """
        logger.info("serper_news", query=query[:50])

        payload = {
            "q": query,
            "num": min(num, 100),
            "gl": gl,
            "hl": hl
        }
        if tbs:
            payload["tbs"] = tbs

        result = await self.post("/news", json=payload)

        return {
            "query": query,
            "news": result.get("news", []),
            "total_results": len(result.get("news", []))
        }

    async def search_images(
        self,
        query: str,
        num: int = 10,
        gl: str = "br"
    ) -> Dict[str, Any]:
        """Busca de imagens"""
        logger.info("serper_images", query=query[:50])

        result = await self.post("/images", json={
            "q": query,
            "num": min(num, 100),
            "gl": gl
        })

        return {
            "query": query,
            "images": result.get("images", []),
            "total_results": len(result.get("images", []))
        }

    async def search_places(
        self,
        query: str,
        location: str = "Brazil",
        gl: str = "br"
    ) -> Dict[str, Any]:
        """Busca de lugares (Google Maps)"""
        logger.info("serper_places", query=query[:50])

        result = await self.post("/places", json={
            "q": query,
            "location": location,
            "gl": gl
        })

        return {
            "query": query,
            "places": result.get("places", []),
            "total_results": len(result.get("places", []))
        }

    async def autocomplete(
        self,
        query: str,
        gl: str = "br",
        hl: str = "pt-br"
    ) -> List[str]:
        """Google Autocomplete"""
        result = await self.post("/autocomplete", json={
            "q": query,
            "gl": gl,
            "hl": hl
        })

        return result.get("suggestions", [])

    # ===========================================
    # MÉTODOS ESPECÍFICOS PARA EMPRESAS
    # ===========================================

    async def find_company_cnpj(self, company_name: str) -> Optional[str]:
        """
        Busca CNPJ de uma empresa pelo nome

        Args:
            company_name: Nome da empresa

        Returns:
            CNPJ encontrado ou None
        """
        query = f'"{company_name}" CNPJ site:cnpj.info OR site:consultacnpj.com OR site:empresascnpj.com'
        results = await self.search(query, num=5)

        # Regex para CNPJ
        cnpj_pattern = r'\d{2}\.?\d{3}\.?\d{3}/?\d{4}-?\d{2}'

        for item in results.get("organic", []):
            # Buscar no título e snippet
            text = f"{item.get('title', '')} {item.get('snippet', '')}"
            match = re.search(cnpj_pattern, text)
            if match:
                cnpj = "".join(filter(str.isdigit, match.group()))
                logger.info("cnpj_found", company=company_name, cnpj=cnpj[:8] + "****")
                return cnpj

        # Tentar no knowledge graph
        kg = results.get("knowledge_graph", {})
        if kg:
            for key in ["cnpj", "CNPJ", "registration"]:
                if key in kg:
                    cnpj = "".join(filter(str.isdigit, str(kg[key])))
                    if len(cnpj) == 14:
                        return cnpj

        return None

    async def find_company_website(self, company_name: str) -> Optional[str]:
        """
        Busca o website oficial de uma empresa

        Args:
            company_name: Nome da empresa

        Returns:
            URL do website ou None
        """
        query = f'"{company_name}" site oficial'
        results = await self.search(query, num=5)

        for item in results.get("organic", []):
            url = item.get("link", "")
            # Filtrar sites de consulta e priorizar domínios próprios
            if not any(x in url for x in [
                "cnpj.info", "consultacnpj", "empresascnpj",
                "linkedin.com", "facebook.com", "instagram.com",
                "wikipedia.org", "reclameaqui.com"
            ]):
                return url

        return None

    async def find_company_linkedin(self, company_name: str) -> Optional[str]:
        """
        Busca o perfil LinkedIn de uma empresa

        Args:
            company_name: Nome da empresa

        Returns:
            URL do LinkedIn ou None
        """
        query = f'"{company_name}" site:linkedin.com/company'
        results = await self.search(query, num=3)

        for item in results.get("organic", []):
            url = item.get("link", "")
            if "linkedin.com/company" in url:
                return url

        return None

    async def find_company_info(self, company_name: str) -> Dict[str, Any]:
        """
        Busca informações completas sobre uma empresa

        Args:
            company_name: Nome da empresa

        Returns:
            Dicionário com informações encontradas
        """
        logger.info("serper_company_info", company=company_name)

        # Busca principal
        main_results = await self.search(f'"{company_name}" empresa Brasil', num=10)

        # Busca de notícias
        news_results = await self.search_news(f'"{company_name}"', num=5)

        # Extrair knowledge graph se disponível
        kg = main_results.get("knowledge_graph", {})

        return {
            "company_name": company_name,
            "search_results": main_results.get("organic", []),
            "knowledge_graph": kg,
            "news": news_results.get("news", []),
            "website": kg.get("website") or await self.find_company_website(company_name),
            "description": kg.get("description"),
            "industry": kg.get("industry") or kg.get("type"),
            "founded": kg.get("founded") or kg.get("foundingDate"),
            "headquarters": kg.get("headquarters") or kg.get("address"),
            "employees": kg.get("employees") or kg.get("numberOfEmployees"),
            "related_searches": main_results.get("related_searches", [])
        }

    # ===========================================
    # MÉTODOS ESPECÍFICOS PARA PESSOAS
    # ===========================================

    async def find_person_linkedin(self, name: str, company: Optional[str] = None) -> Optional[str]:
        """
        Busca o perfil LinkedIn de uma pessoa

        Args:
            name: Nome da pessoa
            company: Empresa (opcional, melhora precisão)

        Returns:
            URL do LinkedIn ou None
        """
        query = f'"{name}" site:linkedin.com/in'
        if company:
            query += f' "{company}"'

        results = await self.search(query, num=5)

        for item in results.get("organic", []):
            url = item.get("link", "")
            if "linkedin.com/in/" in url:
                return url

        return None

    async def find_person_info(
        self,
        name: str,
        context: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Busca informações sobre uma pessoa

        Args:
            name: Nome da pessoa
            context: Contexto adicional (empresa, cargo, cidade)

        Returns:
            Informações encontradas
        """
        query = f'"{name}"'
        if context:
            query += f' {context}'

        results = await self.search(query, num=10)
        news = await self.search_news(f'"{name}"', num=5)

        return {
            "name": name,
            "search_results": results.get("organic", []),
            "knowledge_graph": results.get("knowledge_graph"),
            "news": news.get("news", []),
            "linkedin": await self.find_person_linkedin(name),
            "people_also_ask": results.get("people_also_ask", [])
        }

    # ===========================================
    # MÉTODOS ESPECÍFICOS PARA POLÍTICOS
    # ===========================================

    async def find_politician_info(
        self,
        name: str,
        role: Optional[str] = None,
        state: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Busca informações sobre um político

        Args:
            name: Nome do político
            role: Cargo (prefeito, senador, deputado, etc)
            state: Estado (UF)

        Returns:
            Informações encontradas
        """
        query_parts = [f'"{name}"']
        if role:
            query_parts.append(role)
        if state:
            query_parts.append(state)
        query_parts.append("político")

        query = " ".join(query_parts)

        results = await self.search(query, num=15)
        news = await self.search_news(f'"{name}" político', num=10)

        # Busca específica em sites de transparência
        gov_query = f'"{name}" site:gov.br OR site:camara.leg.br OR site:senado.leg.br'
        gov_results = await self.search(gov_query, num=5)

        return {
            "name": name,
            "role": role,
            "state": state,
            "search_results": results.get("organic", []),
            "knowledge_graph": results.get("knowledge_graph"),
            "news": news.get("news", []),
            "gov_results": gov_results.get("organic", []),
            "social_media": {
                "instagram": await self._find_social(name, "instagram.com"),
                "twitter": await self._find_social(name, "twitter.com"),
                "facebook": await self._find_social(name, "facebook.com")
            }
        }

    async def _find_social(self, name: str, domain: str) -> Optional[str]:
        """Busca perfil em rede social"""
        query = f'"{name}" site:{domain}'
        results = await self.search(query, num=1)

        organic = results.get("organic", [])
        if organic:
            return organic[0].get("link")
        return None

File: backend/src/services/serper.js
const SERPER_API_KEY = process.env.SERPER_API_KEY;
const SERPER_BASE_URL = 'https://google.serper.dev';

/**
 * Make request to Serper API
 * @param {string} endpoint - API endpoint
 * @param {Object} payload - Request payload
 * @returns {Promise<Object>} API response
 */
async function serperRequest(endpoint, payload) {
  const response = await fetch(`${SERPER_BASE_URL}${endpoint}`, {
    method: 'POST',
    headers: {
      'X-API-KEY': SERPER_API_KEY,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(payload)
  });

  if (!response.ok) {
    throw new Error(`Serper API error: ${response.status}`);
  }

  return response.json();
}

/**
 * Search Google via Serper
 * @param {string} query - Search query
 * @param {number} num - Number of results
 * @returns {Promise<Object>} Search results
 */
export async function search(query, num = 10) {
  return serperRequest('/search', {
    q: query,
    num: Math.min(num, 100),
    gl: 'br',
    hl: 'pt-br'
  });
}

/**
 * Search for company by name, find CNPJ candidates
 * @param {string} companyName - Company name to search
 * @returns {Promise<Array>} List of company candidates with CNPJ
 */
export async function searchCompanyByName(companyName) {
  // Search in CNPJ databases
  const query = `"${companyName}" CNPJ site:cnpj.info OR site:consultacnpj.com OR site:casadosdados.com.br`;
  const results = await search(query, 20);

  const candidates = [];
  const cnpjPattern = /\d{2}\.?\d{3}\.?\d{3}\/?\d{4}-?\d{2}/g;
  const seenCnpjs = new Set();

  // Extract from organic results
  for (const item of results.organic || []) {
    const text = `${item.title || ''} ${item.snippet || ''}`;
    const matches = text.match(cnpjPattern) || [];

    for (const match of matches) {
      const cnpj = match.replace(/[^\d]/g, '');
      if (cnpj.length === 14 && !seenCnpjs.has(cnpj)) {
        seenCnpjs.add(cnpj);

        // Extract location from snippet
        const locationMatch = item.snippet?.match(/([A-Z]{2})\s*[-–]\s*([^,\n]+)/i);

        candidates.push({
          cnpj: cnpj,
          cnpj_formatted: formatCnpj(cnpj),
          razao_social: extractCompanyName(item.title, item.snippet),
          localizacao: locationMatch ? `${locationMatch[2].trim()} - ${locationMatch[1]}` : null,
          fonte_url: item.link,
          snippet: item.snippet
        });
      }
    }
  }

  // Also search in knowledge graph
  const kg = results.knowledgeGraph;
  if (kg) {
    const cnpjFromKg = kg.cnpj || kg.CNPJ;
    if (cnpjFromKg) {
      const cnpj = cnpjFromKg.replace(/[^\d]/g, '');
      if (cnpj.length === 14 && !seenCnpjs.has(cnpj)) {
        candidates.unshift({
          cnpj: cnpj,
          cnpj_formatted: formatCnpj(cnpj),
          razao_social: kg.title || companyName,
          localizacao: kg.headquarters || kg.address,
          fonte_url: kg.website,
          snippet: kg.description,
          knowledge_graph: true
        });
      }
    }
  }

  return candidates;
}

/**
 * Get detailed company info by CNPJ
 * @param {string} cnpj - Company CNPJ
 * @returns {Promise<Object>} Company details
 */
export async function getCompanyDetails(cnpj) {
  const cnpjFormatted = formatCnpj(cnpj);

  // Search for company details
  const query = `CNPJ ${cnpjFormatted}`;
  const results = await search(query, 10);

  // Search for LinkedIn
  const linkedinQuery = `CNPJ ${cnpjFormatted} site:linkedin.com/company`;
  const linkedinResults = await search(linkedinQuery, 3);

  let linkedin = null;
  for (const item of linkedinResults.organic || []) {
    if (item.link?.includes('linkedin.com/company')) {
      linkedin = item.link;
      break;
    }
  }

  // Extract info from knowledge graph
  const kg = results.knowledgeGraph || {};

  return {
    cnpj: cnpj,
    cnpj_formatted: cnpjFormatted,
    razao_social: kg.title || null,
    nome_fantasia: null,
    website: kg.website || null,
    linkedin: linkedin,
    endereco: kg.address || kg.headquarters || null,
    cidade: null,
    estado: null,
    setor: kg.industry || kg.type || null,
    descricao: kg.description || null,
    data_fundacao: kg.founded || kg.foundingDate || null,
    num_funcionarios: kg.employees || kg.numberOfEmployees || null,
    fundadores: extractFounders(results),
    organic_results: results.organic || []
  };
}

/**
 * Extract founders from search results
 * @param {Object} results - Search results
 * @returns {Array} List of founders
 */
function extractFounders(results) {
  const founders = [];
  const kg = results.knowledgeGraph || {};

  // Check for founders in knowledge graph
  if (kg.founders) {
    const foundersList = Array.isArray(kg.founders) ? kg.founders : [kg.founders];
    for (const founder of foundersList) {
      founders.push({
        nome: typeof founder === 'string' ? founder : founder.name,
        linkedin: null
      });
    }
  }

  // Check for CEO/executives
  if (kg.ceo) {
    founders.push({
      nome: typeof kg.ceo === 'string' ? kg.ceo : kg.ceo.name,
      cargo: 'CEO',
      linkedin: null
    });
  }

  return founders;
}

/**
 * Search LinkedIn for a person
 * @param {string} name - Person name
 * @param {string} company - Company name (optional)
 * @returns {Promise<string|null>} LinkedIn URL or null
 */
export async function findPersonLinkedin(name, company = null) {
  let query = `"${name}" site:linkedin.com/in`;
  if (company) {
    query += ` "${company}"`;
  }

  const results = await search(query, 5);

  for (const item of results.organic || []) {
    if (item.link?.includes('linkedin.com/in/')) {
      return item.link;
    }
  }

  return null;
}

/**
 * Format CNPJ with punctuation
 * @param {string} cnpj - CNPJ digits only
 * @returns {string} Formatted CNPJ
 */
function formatCnpj(cnpj) {
  const digits = cnpj.replace(/[^\d]/g, '');
  if (digits.length !== 14) return cnpj;
  return `${digits.slice(0,2)}.${digits.slice(2,5)}.${digits.slice(5,8)}/${digits.slice(8,12)}-${digits.slice(12)}`;
}

/**
 * Extract company name from title/snippet
 * @param {string} title - Result title
 * @param {string} snippet - Result snippet
 * @returns {string|null} Company name
 */
function extractCompanyName(title, snippet) {
  // Try to extract from title (usually format: "COMPANY NAME - CNPJ...")
  if (title) {
    const match = title.match(/^([^-–|]+)/);
    if (match) {
      return match[1].trim();
    }
  }
  return title || null;
}
